pallets/dapp-staking/src/lib.rs

```diff
@@ -1516,9 +1336,363 @@ pub mod pallet {
+                            staking_info.unstake(amount, current_era, protocol_state.subperiod());
+
+                        (
+                            staking_info,
+                            amount,
+                            stake_amount_iter,
+                            updated_bonus_status,
+                        )
+                    }
+                    None => {
+                        return Err(Error::<T>::NoStakingInfo.into());
+                    }
+                };
+
+            // 2.
+            // Reduce stake amount
+            ledger
+                .unstake_amount(amount, current_era, protocol_state.period_info)
+                .map_err(|err| match err {
+                    AccountLedgerError::InvalidPeriod | AccountLedgerError::InvalidEra => {
+                        Error::<T>::UnclaimedRewards
+                    }
+                    // This is a defensive check, which should never happen since we calculate the correct value above.
+                    AccountLedgerError::UnstakeAmountLargerThanStake => {
+                        Error::<T>::UnstakeAmountTooLarge
+                    }
+                    _ => Error::<T>::InternalUnstakeError,
+                })?;
+
+            // 3.
+            // Update `ContractStake` storage with the reduced stake amount on the specified contract.
+            let mut contract_stake_info = ContractStake::<T>::get(&dapp_info.id);
+            contract_stake_info.unstake(
+                &stake_amount_iter,
+                protocol_state.period_info,
+                current_era,
+            );
+
+            // 4.
+            // Update total staked amount for the next era.
+            CurrentEraInfo::<T>::mutate(|era_info| {
+                era_info.unstake_amount(amount);
+            });
+
+            // 5.
+            // Update remaining storage entries
+            ContractStake::<T>::insert(&dapp_info.id, contract_stake_info);
+
+            if new_staking_info.is_empty() {
+                ledger.contract_stake_count.saturating_dec();
+                StakerInfo::<T>::remove(&account, &smart_contract);
+            } else {
+                StakerInfo::<T>::insert(&account, &smart_contract, new_staking_info);
+            }
+
+            Self::update_ledger(&account, ledger)?;
+
+            // Implementation comment, remove once production code is ready:
+            // We need to know what & how much was unstaked. The code can be made cleaner later, maybe there are redundant parts now.
+            // Return the `StakeAmount` that has max total value - that's the one that is equal to the `amount` parameter.
+            let unstake_amount = stake_amount_iter
+                .iter()
+                .max_by(|a, b| a.compare_stake_amounts(b))
+                .expect("At least one value exists, otherwise we wouldn't be here.");
+            assert_eq!(unstake_amount.total(), amount);
```

Comment:
I've left this comment in another place, but let me repeat the question here too - why is this custom comparison needed? Why isn't `max total value` good enough?

My understanding is that:
* amounts provided in the iterator will always refer to two consecutive eras
* _future_ era will **always** have a higher unstaked amount than the current era (is this correct?)
  * this is because we can never have staked more in the current era, than in the future era
  * stake operation always increases stake amount for the _next_era
  * unstake operation first reduces future era stake, then current era stake, aligning the values

So taking the simple `total max`, gives the amount that was unstaked.

What scenario am I not seeing?

Reply:
Using `current_era` in the method footprints solved it